setup.py

```diff
@@ -1,115 +1,54 @@
-"""Grader: a Python module and command-line utility to grade applications.
-
-This setup.py is based on the template at the PyPA sample project. See:
-https://packaging.python.org/en/latest/distributing.html
-https://github.com/pypa/sampleproject
-"""
-
-# Always prefer setuptools over distutils
+# -*- coding: utf-8 -*-
+import ast
+import os
 from setuptools import setup, find_packages
-# To use a consistent encoding
-from codecs import open
-from os import path
 
-here = path.abspath(path.dirname(__file__))
+# this file is used to pick the relevant metadata for setup.py
+INITFILE = os.path.join('grader', '__init__.py')
+# the directory we are in
+CWD = os.path.abspath(os.path.dirname(__file__))
+
+def parse_keyword(key):
+    """Get metadata from grader/__init__.py using an AST"""
+    with open(os.path.join(CWD, INITFILE), encoding='utf-8') as f:
+        ast_tree = ast.parse(f.read())
+        for node in ast.walk(ast_tree):
+            if type(node) is ast.Assign:
+                try:
+                    if node.targets[0].id == key:
+                        return node.value.s
+                except:
+                    pass
+    # return "not available" if we didn't find the variable
+    return 'N/A'
+
+# pick the relevant keywords from the __init__.py file
+metadata_vars = ('name', 'version', 'description', 'author', 'license', 'url')
+metadata = dict((var, parse_keyword('__%s__'%var)) for var in metadata_vars)
```

Comment:
This is to avoid `from grader import __version__, __author__, ...` and `setup(version=__version__, author=__author__, ...)` right? 😉

Reply:
A dict comprehension is more readable and concise imho:

```python
metadata = {var: parse_keyword('__%s__' % var) for var in metadata_vars}
```

or, if you want to support only 3.6+:

```python
metadata = {var: parse_keyword(f'__{var}__') for var in metadata_vars}
```